From 78aa50caa9e9a24f0ffc017bbba3857805cd5c36 Mon Sep 17 00:00:00 2001
Message-Id: <78aa50caa9e9a24f0ffc017bbba3857805cd5c36.1533858330.git.osandov@osandov.com>
From: Omar Sandoval <osandov@osandov.com>
Date: Tue, 4 Apr 2017 14:40:27 -0700
Subject: [PATCH 1/2] Whitelist key event modifiers for shortcuts

I'm seeing a weird issue with my xserver where all key press events will
be set with (state & Button1Mask), which ends up breaking all st
keyboard shortcuts. xterm works correctly because it whitelists
modifiers relevant to key press events. Do the same in st.
---
 win.h | 3 +++
 x.c   | 3 ++-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/win.h b/win.h
index 31f327d..503263b 100644
--- a/win.h
+++ b/win.h
@@ -1,5 +1,8 @@
 /* See LICENSE for license details. */
 
+#define KeyModifiers (ShiftMask | LockMask | ControlMask | Mod1Mask | \
+		      Mod2Mask | Mod3Mask | Mod4Mask | Mod5Mask)
+
 enum win_mode {
 	MODE_VISIBLE     = 1 << 0,
 	MODE_FOCUSED     = 1 << 1,
diff --git a/x.c b/x.c
index 00cb6b1..9d13b51 100644
--- a/x.c
+++ b/x.c
@@ -1644,7 +1644,8 @@ focus(XEvent *ev)
 int
 match(uint mask, uint state)
 {
-	return mask == XK_ANY_MOD || mask == (state & ~ignoremod);
+	return (mask == XK_ANY_MOD ||
+		mask == (state & KeyModifiers & ~ignoremod));
 }
 
 char*
-- 
2.18.0

